use WebAssembly.instantiateStreaming

xtuc opened the pull request:
This is probably going to be a long PR, that's just an early preview.

<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**What kind of change does this PR introduce?**

wasm-edit and WebAssembly.instantiateStreaming

Also includes https://github.com/webpack/webpack/pull/6531.

Closes #6433 

For ref https://github.com/xtuc/webassemblyjs/pull/196

<!-- E.g. a bugfix, feature, refactoring, build related change, etc… -->

**Did you add tests for your changes?**

:+1: Not needed, existing tests.

<!-- Note that we won't merge your changes if you don't add tests -->

**If relevant, link to documentation update:**

:+1: nothing changes from the user point of view

<!-- Link PR from webpack/webpack.js.org here, or N/A -->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**

:+1: no, it mostly unbreaks

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

### A few open thoughts 

- Node 8 is not required anymore (because in https://github.com/webpack/webpack/pull/6531 it does)
- Currently each edit/add/decode decodes the binary, I will find a way to batch them and avoid the unnecessary work

---

Unfortunately using a WebAssembly.Table for the interop with Webpack don't work because we are not able to `set` JavaScript functions.

The table has some good advantages tho:
  - avoids shifting every call indices (because it uses call_indirect instead)
  - reduces the integration points webpack<>wasm
  - will hold more type of values in the future which would fix the live-binding issue
  - when multiple table will be allowed that will be completely isolated

--- 

Another idea, we could inject a new function that will fill the globals that were imported.

For example:
```wat
(module
  (import "env" "global" (global i32))
  (func (export "t") (result i32)
    (get_global 0)
  )
)
```

Transformations:
- remove the global import (because you can't import a mutable global)
- adds the init function

```wat
(module
  (global (mut i32) (i32.const 0))
  (func (export "t") (result i32)
    (get_global 0)
  )
  (func (export "init") (param i32)
    (get_local 0)
    (set_global 0)
  )
)
```

We first call `init` with the value of the global, which will then call the start function if provided.

This solution doesn't require much transformation to the binary. According to v8's limits, a function can have 1k params (so 1k global imports, note that if that doesn't suffice we can add a second init function): 
https://github.com/v8/v8/blob/db1e370decbd4d2acd4b5c2ebfee76b3c24c5cae/src/wasm/wasm-limits.h#L36

While we change the mutability of the global, that program has already been compiled and ensures that the user don't modify the global (I hope).

**Globals and functions indices are preserved**.

xtuc commented on examples/sven/src/index.js:
```diff
@@ -0,0 +1,22 @@
+import("./test.wasm").then(({getTwo, logFoo}) => {
+  console.log("getTwo", getTwo());
+  console.log(logFoo());
+})
```
Here's a small example.

xtuc commented on lib/wasm/WasmModuleTemplatePlugin.js:
```diff
@@ -40,50 +40,11 @@ class WasmModuleTemplatePlugin {
-							"var instance = new WebAssembly.Instance(__webpack_require__.w[module.i], {" +
-								generateImports(),
-							"});",
+							"var instance = __webpack_require__.w[module.i]",
```
I made the choose that `__webpack_require__.w` only contains the `WebAssembly.instance` (`WebAssembly.Module` is not preserve because not used)

xtuc commented on lib/web/FetchCompileWasmMainTemplatePlugin.js:
```diff
@@ -61,6 +142,10 @@ class FetchCompileWasmMainTemplatePlugin {
 					"// Fetch + compile chunk loading for webassembly",
 					"",
+					"var importObjects = {",
+					Template.indent([importObjects]),
```
This looks like this:
```js
/******/ 		var importObjects = {
/******/ 			"./src/test.wasm": {
/******/ 				"./b": {
/******/ 					"logFoo": function() {
/******/ 			                                  return __webpack_require__("./src/b.js")["logFoo"]();
/******/ 			                                }
/******/ 				},
/******/ 				"./a": {
/******/ 					"_global_get_two": function() {
/******/ 			                                  return __webpack_require__("./src/a.js")["two"];
/******/ 			                                }
/******/ 				}
/******/ 			}
/******/ 		}
```

xtuc commented:
Unfortunately using a WebAssembly.Table for the interop with Webpack don't work because we are not able to `set` JavaScript functions.

The table has some good advantages tho:
  - avoids shifting every call indices (because it uses call_indirect instead)
  - reduces the integration points webpack<>wasm
  - will hold more type of values in the future which would fix the live-binding issue
  - when multiple table will be allowed that will be completely isolated

xtuc commented:
Another idea, we could inject a new function that will fill the global that were imported.

For example:
```wat
(module
  (import "env" "global" (global i32))
  (func (export "t") (result i32)
    (get_global 0)
  )
)
```

Transformations:
- remove the global import (because you can't import a mutable global)
- adds a init function (which will then call the start function)

```wat
(module
  (global (mut i32) (i32.const 0))
  (func (export "t") (result i32)
    (get_global 0)
  )
  (func (export "init") (param i32)
    (get_local 0)
    (set_global 0)
  )
)
```

We first call `init` with the value of the global, which will then call the start function if provided.

This solution doesn't require much transformation to the binary. According to v8's limits, a function can have 1k params (so 1k global imports, note that if that doesn't suffice we can add a second init function): 
https://github.com/v8/v8/blob/db1e370decbd4d2acd4b5c2ebfee76b3c24c5cae/src/wasm/wasm-limits.h#L36

While we change the mutability of the global, that program has already been compiled and ensures that the user don't modify the global (I hope).

xtuc commented:
@sokra i'm not sure how I could fix this error?

The code is here https://github.com/webpack/webpack/pull/6709/files#diff-c5507c5a9d1837a64e829e10cdd9348bR73

It seems that the instance is not instantiated when the test runs.

```js
  102) TestCases all-combined wasm simple should allow to run a WebAssembly module (direct):
     TypeError: Cannot read property '__init__' of undefined
      at Object../wasm/simple/wasm.wasm?2 (/x/xtuc/webpack/test/js/all-combined/wasm/simple/0.bundle.js:1:168)
      at s (/x/xtuc/webpack/test/js/all-combined/wasm/simple/bundle.js:1:6756)
      at r (/x/xtuc/webpack/test/js/all-combined/wasm/simple/bundle.js:1:729)
      at <anonymous>:null:null
```

sokra commented on lib/web/FetchCompileWasmMainTemplatePlugin.js:
```diff
@@ -26,6 +27,88 @@ class FetchCompileWasmMainTemplatePlugin {
+								)})[${JSON.stringify(data.usedName)}](${params})`;
+							}
+
+							if (data.description.type === "GlobalType") {
```
Is this used? I guess all globals are filtered by the first loop.

xtuc replied:
Yes, I removed it

sokra commented on lib/web/FetchCompileWasmMainTemplatePlugin.js:
```diff
@@ -70,27 +157,26 @@ class FetchCompileWasmMainTemplatePlugin {
-								]),
-								"}"
+								// FIXME(sven): ensrue this still works / change it
+								"promises.push(response.arrayBuffer().then(function(bytes) { installedWasmModules[wasmModuleId] = WebAssembly.compile(bytes); }));"
```
`WebAssembly.compile` -> `WebAssembly.instantiate`

It returns a promise and the result need to be stored in installedWasmModules

xtuc replied:
Yeah I didn't change this branch, that's why I added the FIXME. I'll fix this at the end so I can also test it

xtuc commented:
@sokra update:
- [x] Node and Web uses the same base class. You have different entry method: `applyWeb` and `applyNode`.
- [x] Fixed when `WebAssembly.insantiateStreaming` is not available
- [x] Fixed on Node
- [ ] remove the node restriction in the tests (assuming there's one) since with webassemblyjs/wasm-parser doesn't require a specific version
- [ ] it seems that some stats tests are failing (can't repro locally for some reason)
- [ ] the transformation needs a bit a work still to pass all the tests

sokra commented on lib/BaseWasmMainTemplatePlugin.js:
```diff
@@ -0,0 +1,242 @@
+		this.apply(mainTemplate, generateLoadBinaryCode);
+	}
+
+	applyNode(mainTemplate) {
```
It's a bit weird to mix both targets into one file. Instead move `applyNode` to `ReadFileCompileWasmMainTemplatePlugin.apply` (`ReadFileCompileWasmMainTemplatePlugin extends BaseWasmMainTemplatePlugin` and resp. for `applyWeb`

sokra replied:
Instead of different `apply` methods you can also make `generateLoadBinaryCode` an abstract method.

xtuc replied:
Ok, I can refactor to use some class inheritance or I can just move the specific code from the base to the callers? What's the best to you?

xtuc replied:
I'm more for "composition over inheritance", I changed in https://github.com/webpack/webpack/pull/6709/commits/405fea671972a9920e1aed711758a5a1e051d5b1

xtuc commented:
The failing stats tests are because the they are run against master. I'll merge master here at some point to fix them.

xtuc commented:
lgtm, could you please take a look @sokra?

I'll do followup PRs for memory and table support.

xtuc commented on test/cases/wasm/import-memory/env.js:
```diff
@@ -0,0 +1 @@
+export const mem = new WebAssembly.Memory({ initial: 1 });
```
This Memory object is actually not used! Webpack will automatically adds a new Memory object in the ImportObject.

In the future we will be able to use the user's provided one.

xtuc commented on lib/WebAssemblyParser.js:
```diff
@@ -17,37 +60,41 @@ class WebAssemblyParser extends Tapable {
+				state.module.addDependency(dep);
+
+				// To avoid unexpected behaviors we check that the Memory or Table
+				// has been exported and is not accessed by the user.
```
I would like to warn the error when we have unsupported stuff. This is of course not the proper way of doing it but it's just an idea for now. What do you think?

Could you please indicate me how I could parse and traverse the module?

sokra commented on lib/node/ReadFileCompileWasmTemplatePlugin.js:
```diff
@@ -4,16 +4,40 @@
+
+                      try {
+                        readFile(join(__dirname, ${
+													path
```
spacing is weird here

sokra commented on lib/wasm/WasmModuleTemplatePlugin.js:
```diff
@@ -7,6 +7,27 @@
+		if (dep instanceof WebAssemblyImportDependency) {
+			if (dep.description.type === "GlobalType") {
+				const exportName = dep.name;
+				const usedName = dep.module && dep.module.isUsed(exportName);
```
could be falsy

ashleygwilliams commented:
hello! i am a member of the Rust-WASM WG and we would really love to see this PR merged. is there anything we can do to help get it over the finish line?

sokra commented:
Hi @ashleygwilliams 

I want to merge @xtuc code in a way keeping the old mode and allowing to choose between modes (i. e. when unsupported fetures are used, or when running in other environments like webworker or node.js).

I hope to be able to takle that this week.

xtuc commented:
@sokra feel free to ping me, I can make some time to help you.

sokra commented:
merged into next branch

sokra commented:
see #7144
